dedupe.js

```diff
@@ -1,100 +1,71 @@
+ http://jedwatson.github.io/classnames
+ */
 /* global define */
 
 (function () {
 	'use strict';
 
-	var classNames = (function () {
-		// don't inherit from Object so we can skip hasOwnProperty check later
-		// http://stackoverflow.com/questions/15518328/creating-js-object-with-object-createnull#answer-21079232
-		function StorageObject() {}
-		StorageObject.prototype = Object.create(null);
-
-		function _parseArray (resultSet, array) {
-			var length = array.length;
-
-			for (var i = 0; i < length; ++i) {
-				_parse(resultSet, array[i]);
+	var hasOwn = Object.prototype.hasOwnProperty;
+	var isArray = Array.isArray;
+	var SPACE = /\s+/;
+	
+	// don't inherit from Object so we can skip hasOwnProperty check later
+	// http://stackoverflow.com/questions/15518328/creating-js-object-with-object-createnull#answer-21079232
+	function StorageObject() {}
+	StorageObject.prototype = Object.create(null);
+
+	function _parse (resultSet, arg) {
+		if (!arg) return;
+		
+		var i, length;
+		
+		if (typeof arg === 'string') { // 'foo bar'
+			var array = arg.split(SPACE);
+			length = array.length;
+
+			for (i = 0; i < length; ++i) {
+				resultSet[array[i]] = true;
 			}
-		}
-
-		var hasOwn = {}.hasOwnProperty;
-
-		function _parseNumber (resultSet, num) {
-			resultSet[num] = true;
-		}
+		} else if (isArray(arg)) { // ['foo', 'bar', ...]
+			length = arg.length;
 
-		function _parseObject (resultSet, object) {
-			for (var k in object) {
-				if (hasOwn.call(object, k)) {
+			for (i = 0; i < length; ++i) {
+				_parse(resultSet, arg[i]);
+			}
+		} else if (typeof arg === 'object') { // { 'foo': true, ... }
+			for (i in arg) {
+				if (hasOwn.call(arg, i)) {
 					// set value to false instead of deleting it to avoid changing object structure
 					// https://www.smashingmagazine.com/2012/11/writing-fast-memory-efficient-javascript/#de-referencing-misconceptions
-					resultSet[k] = !!object[k];
+					resultSet[i] = !!arg[i];
 				}
 			}
+		} else if (typeof arg === 'number') { // '130'
+			resultSet[arg] = true;
 		}
+	}
+	
+	function classNames(){
+		var classSet = new StorageObject();
 
-		var SPACE = /\s+/;
-		function _parseString (resultSet, str) {
-			var array = str.split(SPACE);
-			var length = array.length;
-
-			for (var i = 0; i < length; ++i) {
-				resultSet[array[i]] = true;
-			}
-		}
-
-		function _parse (resultSet, arg) {
-			if (!arg) return;
-			var argType = typeof arg;
-
-			// 'foo bar'
-			if (argType === 'string') {
-				_parseString(resultSet, arg);
-
-			// ['foo', 'bar', ...]
-			} else if (Array.isArray(arg)) {
-				_parseArray(resultSet, arg);
-
-			// { 'foo': true, ... }
-			} else if (argType === 'object') {
-				_parseObject(resultSet, arg);
+		var str = "", n,
+			len = arguments.length;
 
-			// '130'
-			} else if (argType === 'number') {
-				_parseNumber(resultSet, arg);
-			}
+		for (n = 0; n < len; n++) {
+			_parse(classSet, arguments[n]);
 		}
 
-		function _classNames () {
```

Comment:
@JedWatson almost the same optimisation reason here (but the real showstopper here is strings split, with which we can do nothing).
But here we had even worse situation in terms of iteration:
- Lines 76-80 are literally useless. It creates new, real array of `arguments` elemnts and then araing iterates over newborn array at lines 87-91; It would've made sence if you passes that array further to other functions but not here.
- And then, again, simply aviding the array accumulator while building the final string;
- Futrthermore i've simply merged parsers into single function, that gives us economy on function calls;